src/rgw/rgw_user.cc

```diff
@@ -1936,6 +1936,216 @@ int RGWUser::check_op(RGWUserAdminOpState& op_state, std::string *err_msg)
+    std::map<std::string, RGWBucketEnt>::iterator it;
+
+    for (it = m.begin(); it != m.end(); ++it) {
+      RGWBucketEnt obj = it->second;
+      ret = rgw_unlink_bucket(store, old_uid, obj.bucket.tenant, obj.bucket.name);
+      if (ret < 0) {
+        set_err_msg(err_msg, "error unlinking bucket " + cpp_strerror(-ret));
+        return ret;
+      }
+
+      marker = it->first;
+
+      RGWBucketInfo bucket_info;
+      RGWSysObjectCtx sys_ctx = store->svc.sysobj->init_obj_ctx();
+      
+      ret = store->get_bucket_info(sys_ctx, obj.bucket.tenant, obj.bucket.name,
+      bucket_info, NULL, null_yield, &attrs);
+      if (ret < 0) {
+        set_err_msg(err_msg, "failed to fetch bucket info for bucket= " + obj.bucket.name);
+        return ret;
+      }
+
+      RGWObjVersionTracker objv_tracker;
+      RGWObjVersionTracker old_version = bucket_info.objv_tracker;
+
+      ret = store->set_bucket_owner(obj.bucket, owner);
+      if (ret < 0) {
+        set_err_msg(err_msg, "failed to set bucket owner: " + cpp_strerror(-ret));
+        return ret;
+      }
+
+      const rgw_pool& root_pool = store->svc.zone->get_zone_params().domain_root;
+      std::string bucket_entry;
+      rgw_make_bucket_entry_name(obj.bucket.tenant, obj.bucket.name, bucket_entry);
+      rgw_raw_obj ob(root_pool, bucket_entry);
+      auto sysobj = sys_ctx.get_obj(ob);
+      rgw_raw_obj obj_bucket_instance;
+
+      store->get_bucket_instance_obj(obj.bucket, obj_bucket_instance);
+      auto inst_sysobj = sys_ctx.get_obj(obj_bucket_instance);
+      ret = inst_sysobj.wop()
+                .set_objv_tracker(&objv_tracker)
+                .write_attr(RGW_ATTR_ACL, aclbl, null_yield);
+      if (ret < 0) {
+        set_err_msg(err_msg, "failed to set new acl on bucket " + obj.bucket.name);
+        return ret;
+      }
+
+      RGWBucketEntryPoint ep;
+      ep.bucket = bucket_info.bucket;
+      ep.owner = uid;
+      ep.creation_time = bucket_info.creation_time;
+      ep.linked = true;
+      map<string, bufferlist> ep_attrs;
+      rgw_ep_info ep_data{ep, ep_attrs};
+
+      ret = rgw_link_bucket(store, uid, bucket_info.bucket,
+              ceph::real_time(), true, &ep_data);
+      if (ret < 0) {
+        set_err_msg(err_msg, "failed to link bucket " + obj.bucket.name + " to new user");
+        return ret;
+      }
+
+      RGWBucketInfo new_bucket_info;
+      ret = store->get_bucket_info(sys_ctx, obj.bucket.tenant, obj.bucket.name,
+      new_bucket_info, NULL, null_yield, &attrs);
+      if (ret < 0) {
+        set_err_msg(err_msg, "failed to fetch bucket info for bucket= " + obj.bucket.name);
+        return ret;
+      }
+
+      ret = rgw_bucket_chown(store, user_info, new_bucket_info, obj_marker, attrs);
+      if (ret < 0) {
+        set_err_msg(err_msg, "failed to run bucket chown" + cpp_strerror(-ret));
+        return ret;
+      }
+    }
+
+  } while (is_truncated);
+
+  // delete old user
+  ret = execute_remove(op_state, &subprocess_msg);
```

Comment:
i think `execute_remove()` and `rgw_delete_user()` are doing a lot more than we need here; the reason that you need the later call to update() is because rgw_delete_user() is deleting the index objects that were created for the new user in execute_rename()->update()->rgw_store_user_info(). this puts us in a similar situation as before, where execute_remove() is deleting stuff and we could crash/error before it gets recreated by update()

it looks like only the last two parts of rgw_delete_user() are needed here: "removing user buckets index" and "removing user index". can you try replacing execute_remove()+update() with those two calls?

Reply:
Ok, will do that. But what I don't understand is how the new user key is getting deleted. Because I'm clearly separating the namespaces everywhere with no overlaps.

Reply:
the names of these index objects are the same for the old and new user. given a user with `uid=foo email=foo@example.com access-key=ABCDEFG`, there are rados objects `foo@example.com` and `ABCDEFG` that store the uid `foo`. these allow you to look up the user if you only know its email or access-key, using rgw_get_user_info_by_email/by_access_key()

user rename is creating a new user and overwriting those existing index objects to point at the new uid

Reply:
Figured that we weren't getting past the check in L229 of rgw_store_user_info() because of the access_key being the same. I have worked around that. Is there a better way than adding an additional attribute to RGWUserInfo?

Reply:
hmm we shouldn't need to extend RGWUserInfo for this; rgw_store_user_info() can derive it based on old_info:
```c++
const bool renamed = old_info && old_info->user_id != info.user_id;
```

we'd also need to modify the EEXIST checks:
```diff
       int r = rgw_get_user_info_by_access_key(store, k.id, inf);
-      if (r >= 0 && inf.user_id.compare(info.user_id) != 0) {
+      if (r >= 0 && inf.user_id.compare(info.user_id) != 0 &&
+          (!old_info || inf.user_id.compare(old_info->user_id) != 0)) {
```

Reply:
Done